beps/bep_0055.rst

```diff
@@ -0,0 +1,113 @@
+:BEP: 55
+:Title: Holepunch extension
+:Version: $Revision$
+:Last-Modified: $Date$
+:Author:  Mihael Peklar <mihael1peklar@gmail.com>
+:Status:  Draft
+:Type:    Standards Track
+:Requires: 10,29
+:Content-Type: text/x-rst
+:Created: 29-Dec-2018
+:Post-History: 23-Feb-2019: initial version
+
+
+The holepunch extension provides a way to connect to peers that are cannot receive inbound connections, whether they are behind a filtering NAT or a firewall that blocks incoming connections.
+
+
+Protocol Extension
+==================
+
+The holepunch extension introduces a new message ``ut_holepunch`` via the extension protocol [#BEP-10]_.
+
+.. parsed-literal::
+
+    {
+      m: {
+        ut_holepunch: *<implementation-dependent local message ID (positive integer)>*,
+        ...
+      },
+      ...
+    } 
+    
+    
+
+The extension message itself consists of the bittorrent extended message header and the following binary payload:
+
+.. parsed-literal::    
+
+    msg_type (1 byte): *<type of holepunch message>*
+    addr_type (1 byte): *<0x00 for ipv4, 0x01 for ipv6 endpoints>*
+    addr (either 6 or 18 bytes): *<compact-encoded ipv4 or ipv6 endpoint, as determined by addr_type>*
+    err_code (4 bytes): *<error code as a big-endian 4-byte integer; 0 in non-error messages>*
+
+
+
+Message types are defined as follows:
+
+ ======== ============= =====================================================================
+ msg_type name          description
+ ======== ============= =====================================================================
+ 0x00     rendezvous    Send connect messages to both the initiating peer and receiving peer
```

Comment:
"receiving peer" sounds a bit misleading here. It's the peer at the IP specified in the message, not the peer receiving the `rendezvous` message.